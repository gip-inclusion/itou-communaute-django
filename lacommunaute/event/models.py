from django.conf import settings
from django.db import models
from machina.models.abstract_models import DatedModel


class Event(DatedModel):
    name = models.CharField(max_length=100, verbose_name="Nom")
    date = models.DateField(verbose_name="Date de début", help_text="au format JJ/MM/YYYY")
    time = models.TimeField(verbose_name="Heure de début", help_text="au format HH:MM")
    end_date = models.DateField(verbose_name="Date de fin", help_text="au format JJ/MM/YYYY")
    end_time = models.TimeField(verbose_name="Heure de fin", help_text="au format HH:MM")
    color = models.IntegerField(verbose_name="Color", default=1)
    description = models.TextField(verbose_name="Notes", blank=True, null=True)

    location = models.URLField(verbose_name="Lien vers l'évènement", null=True, blank=True)
    poster = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.CASCADE,
        verbose_name="Poster",
    )

    objects = models.Manager()

    class Meta:
        verbose_name = "Event"
        verbose_name_plural = "Events"
        ordering = ["-date", "-time"]

    def __str__(self):
        return f"{self.name} - {self.date}"
