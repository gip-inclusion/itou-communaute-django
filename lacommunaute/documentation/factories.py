import factory
from faker import Faker

from lacommunaute.documentation.models import Category, Document, DocumentRating


faker = Faker()


class AbstractDocumentationFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    description = factory.Faker("sentence", nb_words=100)
    short_description = factory.Faker("sentence", nb_words=10)
    image = factory.django.ImageField(filename="banner.jpg")


class CategoryFactory(AbstractDocumentationFactory):
    class Meta:
        model = Category

    class Params:
        for_snapshot = factory.Trait(
            name="Test Category", description="Test description", short_description="Test short description"
        )


class DocumentFactory(AbstractDocumentationFactory):
    category = factory.SubFactory(CategoryFactory)

    class Meta:
        model = Document

    class Params:
        for_snapshot = factory.Trait(
            name="Test Document",
            description="Test description",
            short_description="Test short description",
            category=factory.SubFactory(CategoryFactory, for_snapshot=True),
        )

    @factory.post_generation
    def with_tags(self, create, extracted, **kwargs):
        if not create or not extracted:
            return

        if isinstance(extracted, list):
            for tag in extracted:
                self.tags.add(tag)


class DocumentRatingFactory(factory.django.DjangoModelFactory):
    document = factory.SubFactory(DocumentFactory)
    rating = factory.Faker("random_int", min=1, max=5)

    class Meta:
        model = DocumentRating
        skip_postgeneration_save = True

    @factory.post_generation
    def set_created(self, create, extracted, **kwargs):
        if extracted:
            self.created = extracted
            self.save()
