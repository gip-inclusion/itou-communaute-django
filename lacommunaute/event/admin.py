from django.contrib import admin

from lacommunaute.event.models import Event


class EventAdmin(admin.ModelAdmin):
    raw_id_fields = ("poster",)


admin.site.register(Event, EventAdmin)
