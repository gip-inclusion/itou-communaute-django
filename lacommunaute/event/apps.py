from django.apps import AppConfig


class ForumUpVoteAppConfig(AppConfig):
    label = "event"
    name = "lacommunaute.event"
    verbose_name = "Event"
    verbose_name_plural = "Events"
