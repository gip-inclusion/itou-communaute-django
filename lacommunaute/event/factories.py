import datetime

import factory
from factory import fuzzy
from faker import Faker

from lacommunaute.event.models import Event
from lacommunaute.users.factories import UserFactory


faker = Faker()


class EventFactory(factory.django.DjangoModelFactory):
    poster = factory.SubFactory(UserFactory)
    name = factory.Faker("name")
    date = fuzzy.FuzzyDate(datetime.date(2023, 5, 17), datetime.date(2023, 7, 16))
    time = datetime.time(10)
    end_date = factory.LazyAttribute(lambda o: o.date)
    end_time = factory.LazyAttribute(lambda o: o.time.replace(hour=(o.time.hour + 1) % 24))
    location = faker.url()
    description = factory.LazyAttribute(lambda t: faker.text(max_nb_chars=200))

    class Meta:
        model = Event
