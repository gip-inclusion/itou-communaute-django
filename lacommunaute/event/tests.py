from dateutil.relativedelta import relativedelta
from django.db import IntegrityError
from django.test import TestCase
from django.utils import timezone

from lacommunaute.event.factories import EventFactory
from lacommunaute.event.models import Event


class EventCurrentUpcomingManagerTest(TestCase):
    def test_manager(self):
        now = timezone.now()

        EventFactory(date=now)

        old_event = EventFactory(date=now - relativedelta(months=1))

        # upcomming event
        EventFactory(date=now + relativedelta(months=1))

        # Event at the beginning of the current month
        EventFactory(date=now.replace(now.year, now.month, 1))

        self.assertEqual(Event.current_and_upcomings.count(), 3)
        self.assertNotIn(old_event, Event.current_and_upcomings.all())


class EventModelTest(TestCase):
    def test_user_is_mandatory(self):
        with self.assertRaises(IntegrityError):
            EventFactory(poster=None)
