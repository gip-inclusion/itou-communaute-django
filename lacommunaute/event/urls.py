from django.urls import path

from lacommunaute.event.views import (
    EventCreateView,
    EventDeleteView,
    EventDetailView,
    EventListView,
    EventMonthArchiveView,
    EventUpdateView,
)


app_name = "event"


urlpatterns = [
    path("create/", EventCreateView.as_view(), name="create"),
    path("<int:pk>/", EventDetailView.as_view(), name="detail"),
    path("<int:pk>/update/", EventUpdateView.as_view(), name="update"),
    path("<int:pk>/delete/", EventDeleteView.as_view(), name="delete"),
    path("myevents/", EventListView.as_view(), name="myevents"),
    path("events/<int:year>/<int:month>/", EventMonthArchiveView.as_view(), name="month"),
    path("events/", EventMonthArchiveView.as_view(), name="current"),
]
