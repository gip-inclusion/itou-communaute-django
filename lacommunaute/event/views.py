import logging
from datetime import datetime

from django.contrib.auth.mixins import LoginRequiredMixin
from django.core.exceptions import PermissionDenied
from django.http import Http404
from django.urls import reverse
from django.views.generic.dates import MonthArchiveView
from django.views.generic.detail import DetailView
from django.views.generic.edit import CreateView, DeleteView, UpdateView
from django.views.generic.list import ListView

from lacommunaute.event.forms import EventModelForm
from lacommunaute.event.models import Event


logger = logging.getLogger(__name__)


class SuccessUrlMixin:
    def get_success_url(self):
        return reverse(
            "event:myevents",
        )


class EventCreateView(LoginRequiredMixin, SuccessUrlMixin, CreateView):
    model = Event
    form_class = EventModelForm

    def form_valid(self, form):
        form.instance.poster = self.request.user
        return super().form_valid(form)


class EventUpdateView(LoginRequiredMixin, SuccessUrlMixin, UpdateView):
    model = Event
    form_class = EventModelForm

    def dispatch(self, request, *args, **kwargs):
        obj = self.get_object()
        if obj.poster == self.request.user:
            return super().dispatch(request, *args, **kwargs)
        raise PermissionDenied

    def form_valid(self, form):
        form.instance.poster = self.request.user
        return super().form_valid(form)


class EventDeleteView(LoginRequiredMixin, SuccessUrlMixin, DeleteView):
    model = Event

    def dispatch(self, request, *args, **kwargs):
        obj = self.get_object()
        if obj.poster == self.request.user:
            return super().dispatch(request, *args, **kwargs)
        raise PermissionDenied


class EventListView(LoginRequiredMixin, ListView):
    model = Event

    def get_queryset(self):
        return Event.objects.filter(poster=self.request.user)


class EventDetailView(DetailView):
    model = Event
    template_name = "event/event_detail.html"


class EventMonthArchiveView(MonthArchiveView):
    allow_future = True
    date_field = "date"
    queryset = Event.objects.all()
    month_format = "%m"
    year_format = "%Y"

    def get_month(self):
        try:
            month = super().get_month()
        except Http404:
            month = datetime.now().strftime(self.get_month_format())
        return month

    def get_year(self):
        try:
            year = super().get_year()
        except Http404:
            year = datetime.now().strftime(self.get_year_format())
        return year
