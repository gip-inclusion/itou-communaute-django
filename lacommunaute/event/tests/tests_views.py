import json
from datetime import datetime, timedelta

from dateutil.relativedelta import relativedelta
from django.test import TestCase
from django.urls import reverse
from django.utils import timezone
from faker import Faker

from lacommunaute.event.factories import EventFactory
from lacommunaute.event.forms import EventModelForm
from lacommunaute.event.models import Event
from lacommunaute.users.factories import UserFactory


faker = Faker()


class EventModelFormTest(TestCase):
    def test_required_fields(self):
        form = EventModelForm(data={})
        self.assertFalse(form.is_valid())
        self.assertEqual(form.errors["name"], ["Ce champ est obligatoire."])
        self.assertEqual(form.errors["date"], ["Ce champ est obligatoire."])
        self.assertEqual(form.errors["time"], ["Ce champ est obligatoire."])
        self.assertEqual(form.errors["end_date"], ["Ce champ est obligatoire."])
        self.assertEqual(form.errors["end_time"], ["Ce champ est obligatoire."])

    def test_dates_are_not_in_the_past(self):
        form = EventModelForm(
            data={
                "date": timezone.now().date() - relativedelta(days=2),
                "time": timezone.now().time(),
                "end_date": timezone.now().date() - relativedelta(days=1),
                "end_time": timezone.now().time(),
            }
        )
        self.assertFalse(form.is_valid())
        self.assertEqual(form.errors["date"], ["La date ne peut pas être dans le passé."])
        self.assertEqual(form.errors["end_date"], ["La date ne peut pas être dans le passé."])

    def test_end_date_is_after_date(self):
        form = EventModelForm(
            data={
                "date": timezone.now().date() + relativedelta(days=1),
                "time": timezone.now().time(),
                "end_date": timezone.now().date(),
                "end_time": timezone.now().time(),
            }
        )
        self.assertFalse(form.is_valid())
        self.assertEqual(form.errors["end_date"], ["La date de fin doit être après la date de début."])

    def test_end_time_is_after_time(self):
        end_time = datetime.now() - timedelta(hours=1)
        form = EventModelForm(
            data={
                "date": datetime.now().date(),
                "time": datetime.now().time(),
                "end_date": datetime.now().date(),
                "end_time": end_time.time(),
            }
        )
        self.assertFalse(form.is_valid())
        self.assertEqual(form.errors["end_time"], ["L'heure de fin doit être après l'heure de début."])


class EventCreateViewTest(TestCase):
    @classmethod
    def setUpTestData(cls):
        cls.user = UserFactory()
        cls.url = reverse("event:create")

    def test_login_is_required(self):
        response = self.client.get(self.url)
        self.assertEqual(response.status_code, 302)
        self.assertEqual(response.url, reverse("inclusion_connect:authorize") + "?next=" + self.url)

    def test_event_is_created(self):
        self.client.force_login(self.user)
        data = {
            "name": faker.name(),
            "date": timezone.now().date(),
            "time": timezone.now().time(),
            "end_date": timezone.now().date(),
            "end_time": timezone.now().time(),
        }
        response = self.client.post(self.url, data)
        self.assertEqual(response.status_code, 302)
        self.assertEqual(Event.objects.count(), 1)
        self.assertEqual(Event.objects.get().poster, self.user)

    def test_event_is_not_created_with_invalid_data(self):
        self.client.force_login(self.user)
        data = {
            "name": faker.name(),
            "date": timezone.now().date(),
            "time": timezone.now().time(),
            "end_date": timezone.now().date() - relativedelta(days=1),
            "end_time": timezone.now().time(),
        }
        response = self.client.post(self.url, data)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(Event.objects.count(), 0)


class EventUpdateViewTest(TestCase):
    @classmethod
    def setUpTestData(cls):
        cls.event = EventFactory()
        cls.user = cls.event.poster
        cls.url = reverse("event:update", kwargs={"pk": cls.event.pk})

    def test_login_is_required(self):
        response = self.client.get(self.url)
        self.assertEqual(response.status_code, 403)

    def test_event_does_not_exist(self):
        self.client.force_login(self.user)
        response = self.client.get(reverse("event:update", kwargs={"pk": 999}))
        self.assertEqual(response.status_code, 404)

    def test_event_is_not_mine(self):
        event = EventFactory()
        self.client.force_login(self.user)
        response = self.client.get(reverse("event:update", kwargs={"pk": event.id}))
        self.assertEqual(response.status_code, 403)

    def test_event_is_mine(self):
        self.client.force_login(self.user)
        response = self.client.get(self.url)
        self.assertEqual(response.status_code, 200)  # 200 = OK

    def test_event_is_updated(self):
        self.client.force_login(self.user)
        response = self.client.get(self.url)
        data = {
            "name": faker.name(),
            "date": timezone.now().date(),
            "time": timezone.now().time(),
            "end_date": timezone.now().date(),
            "end_time": timezone.now().time(),
        }
        response = self.client.post(self.url, data)
        self.assertEqual(response.status_code, 302)
        self.assertEqual(Event.objects.count(), 1)
        event = Event.objects.get()
        self.assertEqual(event.poster, self.user)
        self.assertEqual(event.name, data["name"])
        self.assertEqual(event.date, data["date"])
        self.assertEqual(event.time, data["time"])
        self.assertEqual(event.end_date, data["end_date"])
        self.assertEqual(event.end_time, data["end_time"])


class EventDeleteViewTest(TestCase):
    @classmethod
    def setUpTestData(cls):
        cls.event = EventFactory()
        cls.user = cls.event.poster
        cls.url = reverse("event:delete", kwargs={"pk": cls.event.pk})

    def test_login_is_required(self):
        response = self.client.get(self.url)
        self.assertEqual(response.status_code, 403)

    def test_event_does_not_exist(self):
        self.client.force_login(self.user)
        response = self.client.get(reverse("event:delete", kwargs={"pk": 999}))
        self.assertEqual(response.status_code, 404)

    def test_event_is_not_mine(self):
        event = EventFactory()
        self.client.force_login(self.user)
        response = self.client.get(reverse("event:delete", kwargs={"pk": event.id}))
        self.assertEqual(response.status_code, 403)

    def test_event_is_mine(self):
        self.client.force_login(self.user)
        response = self.client.get(self.url)
        self.assertEqual(response.status_code, 200)


class EventListViewTest(TestCase):
    @classmethod
    def setUpTestData(cls):
        cls.url = reverse("event:myevents")
        cls.user = UserFactory()

    def test_login_is_required(self):
        response = self.client.get(self.url)
        self.assertEqual(response.status_code, 302)
        self.assertEqual(response.url, reverse("inclusion_connect:authorize") + "?next=" + self.url)

        self.client.force_login(self.user)
        response = self.client.get(self.url)
        self.assertEqual(response.status_code, 200)

    def test_list_my_events(self):
        myevent = EventFactory(poster=self.user)
        not_myevent = EventFactory()
        self.client.force_login(self.user)

        response = self.client.get(self.url)
        self.assertEqual(response.status_code, 200)
        self.assertIn(myevent, response.context_data["object_list"])
        self.assertNotIn(not_myevent, response.context_data["object_list"])


class calendar_data_test(TestCase):
    def test_json_response(self):
        event = EventFactory()
        items = [
            {
                "id": event.id,
                "name": event.name,
                "color": 1,
                "location": event.location,
                "description": event.description,
                "poster_id": event.poster.id,
                "time": event.time.strftime("%H:%M:%S"),
                "year": event.date.year,
                "month": event.date.month,
                "day": event.date.day,
                "duration": 1,
            }
        ]
        response = self.client.get(reverse("event:data_source"))
        self.assertEqual(json.loads(response.content)["items"], items)


class calendar_test(TestCase):
    def test_template(self):
        response = self.client.get(reverse("event:calendar"))
        self.assertTemplateUsed(response, "event/event_calendar.html")
        self.assertContains(response, reverse("event:create"))
