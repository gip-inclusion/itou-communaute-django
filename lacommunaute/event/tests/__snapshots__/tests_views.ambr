# serializer version: 1
# name: TestEventMonthArchiveView.test_action_box[False-anonymous][anonymous]
  '''
  <section class="s-section" id="action-box">
          <div class="s-section__container container">
              
                  <div class="s-section__row row">
                      <div class="s-section__col col-12">
                          <div class="c-box mb-5">
                              <div class="row">
                                  <div class="col-lg col-12 mb-3">
                                      Je suis l'organisateur d'un évènement à destination des CIP pour les aider à améliorer leur pratique professionnelle ?
                                      <br/>
                                      Je me connecte pour ajouter un évènement public dans le calendrier de la communauté
                                  </div>
                                  <div class="col-lg-auto col-12">
  
  
      <a class="btn btn-outline-primary btn-ico btn-block" href="/users/login/?next=%2Fcalendar%2Fcreate%2F">Se connecter | S'inscrire</a>
  
  </div>
                              </div>
                          </div>
                      </div>
                  </div>
              
          </div>
      </section>
  '''
# ---
# name: TestEventMonthArchiveView.test_action_box[True-authenticated][authenticated]
  '''
  <section class="s-section" id="action-box">
          <div class="s-section__container container">
              
                  <div class="s-section__row row">
                      <div class="s-section__col col-12">
                          <div class="c-box mb-5">
                              <div class="row">
                                  <div class="col-lg col-12 mb-3">
                                      Je suis l'organisateur d'un évènement à destination des CIP pour les aider à améliorer leur pratique professionnelle ?
                                      <br/>
                                      J'ajoute un évènement public dans le calendrier de la communauté
                                  </div>
                                  <div class="col-lg-auto col-12">
                                      <a class="btn btn-primary" href="/calendar/create/">Ajouter un évènement public</a>
                                  </div>
                              </div>
                          </div>
                      </div>
                  </div>
              
          </div>
      </section>
  '''
# ---
